### PR TITLE
AddingTest Suite for Patient Module

### PR DESCRIPTION
**Pull request by emecas**
Providing Unit Testing for `org.isf.patient.rest.PatientController.java`

- [x] Paths for POST `/patients` API call
- [x] Paths for PUT `/patients/{code}` API call
- [x] Paths for GET `/patients` API call
- [x] Paths for GET `/patients/{code}` API call
- [x] Paths for GET `/patients/search` API call
- [x] Paths for DELETE `/patients/{code}` API call

Note 1: This approach does not required a DB connection or starting the SpringBoot App.

Note 2: There is something I did not like about the multiple calls to `PatientManager` lines 60-61, but I want to provide an initial template for testing all the paths for a particular API call.

Note 3: I had to make protected `DEFAULT_PAGE_SIZE` in `PatientController`

Note 4: I have added `TODO` tasks suggesting changes on API

### CODE REVIEW

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+	/**
+	 * Test method for {@link org.isf.patient.rest.PatientController#deletePatient(int)}.
+	 * @throws Exception 
+	 */
+	@Test
+	public void when_delete_patients_with_existent_code_then_response_true_and_OK() throws Exception {
+		Integer code = 123;
+		String request = "/patients/{code}";
+		PatientDTO expectedPatientDTO =  PatientDTOHelper.setup();
+		expectedPatientDTO.setCode(code);
```
Comment (vicusotto) [inline src/test/java/org/isf/patient/rest/PatientControllerTest.java]: Unused

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+	 * @throws Exception 
+	 */
+	@Test
+	public void when_delete_patients_with_unexistent_code_then_response_Not_Found() throws Exception {
+		Integer code = 111;
+		String request = "/patients/{code}";
+		PatientDTO expectedPatientDTO =  PatientDTOHelper.setup();
+		expectedPatientDTO.setCode(code);
+		Patient	patient = PatientHelper.setupPatient();
+		patient.setCode(code);
```
Comment (vicusotto) [inline src/test/java/org/isf/patient/rest/PatientControllerTest.java]: Unused

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+	/**
+	 * Test method for {@link org.isf.patient.rest.PatientController#deletePatient(int)}.
+	 * @throws Exception 
+	 */
+	@Test
+	public void when_delete_patients_with_existent_code_but_fail_deletion_then_OHAPIException_BadRequest() throws Exception {
+		Integer code = 123;
+		String request = "/patients/{code}";
+		PatientDTO expectedPatientDTO =  PatientDTOHelper.setup();
+		expectedPatientDTO.setCode(code);
```
Comment (vicusotto) [inline src/test/java/org/isf/patient/rest/PatientControllerTest.java]: Unused

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+/**
+ * @author ecastaneda1
+ *
+ */
+
+public class PatientControllerTest {
+	private final Logger logger = LoggerFactory.getLogger(PatientControllerTest.class);
+
+	@Mock
+    private PatientBrowserManager patientBrowserManagerMock = Mockito.mock(PatientBrowserManager.class);
```
[the remark this answers is not in the document]
Comment (emecas) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java]: fixed

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+	 * Test method for {@link org.isf.patient.rest.PatientController#newPatient(org.isf.patient.dto.PatientDTO)}.
+	 * @throws Exception 
+	 */
+	@Test
+	public void when_post_patients_is_call_without_contentType_header_then_HttpMediaTypeNotSupportedException() throws Exception {
+		String request = "/patients";
+		
+		MvcResult result = this.mockMvc
+			.perform(post(request).content(new byte[]{'a', 'b', 'c'}))
+			.andDo(print())
```
[the remark this answers is not in the document]
Comment (emecas) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java]: updated, actually I included it more by my own observation of outputs while writing the tests, but you are totally right.

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+		newPatient.setCode(code);
+		
+		when(patientBrowserManagerMock.newPatient(any(Patient.class))).thenReturn(true);
+		when(patientBrowserManagerMock.getPatient(any(String.class))).thenReturn(newPatient);
+		
+		this.mockMvc
+			.perform(
+					post(request)
+					.contentType(MediaType.APPLICATION_JSON)
+					.content(PatientDTOHelper.asJsonString(newPatientDTO).getBytes()))
```
[the remark this answers is not in the document]
Comment (emecas) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java]: My mistake. Actually, I was using getBytes() in order to cause a BadRequest in another old test and forgot to remove it. it is fixed now.

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+	 * @throws Exception 
+	 */
+	@Test
+	public void when_post_patients_is_call_without_contentType_header_then_HttpMediaTypeNotSupportedException() throws Exception {
+		String request = "/patients";
+		
+		MvcResult result = this.mockMvc
+			.perform(post(request).content(new byte[]{'a', 'b', 'c'}))
+			.andDo(print())
+			.andExpect(status().is4xxClientError())
```
[the remark this answers is not in the document]
Comment (emecas) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java]: yes sure fixed .....  originally I put it more for remembering myself we need adding new exceptions or fixing the returned HTTP status, so the only one without it, it's the one that i would really expect a 401 . 

On the other hand, asserting the 4xx range and then asserting  one explicit 4xx could be some redundant, but for sure could help for fixing/improving all Controllers and the OHResponseEntityExceptionHandler's logic.

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java (hunk 1 above)
Comment (emecas) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java]: Fixed all 3

**Comment by emecas** [comment]
> Your Test structure is good.
> We added some minor feedbacks.

Thank you, I have applied all your reviews

> Some of the Controller APIs are using wrong HTTP status code but this is not an issue related to these tests: they are testing the current PatientController behavior.

Totally agree and I replied with  some comments about it on your reviews, even in the code I kept my original TODO tasks.

> Do you have any idea on how we can test Spring Security?

Personally, I think that testing security is better under integration testing, when you are really running the app using @RunWith(SpringRunner.class) and your app configuration classes.

About testing security, I have used 3 approaches with their pros/cons: 

1. 100% code with TestRestTemplate, RestTemplateBuilder, and a combination the annotations @RunWith(SpringRunner.class) @SpringBootTest &  @MockBean (so you can focus only on testing what you really want to test, in this case security and controller calls)

2. using @RunWith(SpringRunner.class), @Autowired and @WithMockUser method annotation
    - using Spring profiles for injecting specific security Configurations you want to test 
    - or using @WebAppConfiguration  and MockMvcBuilders with .webAppContextSetup(context) and
       .apply(SecurityMockMvcConfigurers.springSecurity())
 
3. using  @RunWith(SpringRunner.class), @SpringBootTest  and the constructor of TestRestTemplate(user, password) for each test with the available exchange methods.

**Comment by mwithi** [comment]
Hi guys, found out that running tests shows warnings about [SLF4J](https://www.baeldung.com/slf4j-classpath-multiple-bindings) bindings. Altough it's just a warning, do you think we can introduce the following exclusions in pom.xml? (it was partially done)
```
<exclusions>
	<exclusion>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-tomcat</artifactId>
	</exclusion>
	<exclusion>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-logging</artifactId>
	</exclusion>
</exclusions>
```
...
```
<dependency>
	<groupId>org.isf</groupId>
	<artifactId>OH-core</artifactId>
	<version>1.9</version>
	<exclusions>
		<exclusion>
			<groupId>org.hibernate.javax.persistence</groupId>
			<artifactId>hibernate-jpa-2.0-api</artifactId>
		</exclusion>
		<exclusion>
			<groupId>org.slf4j</groupId>
			<artifactId>slf4j-log4j12</artifactId>
		</exclusion>
	</exclusions>
</dependency>
```

great work @emecas !

**Comment by emecas** [comment]
> Hi guys, found out that running tests shows warnings about [SLF4J](https://www.baeldung.com/slf4j-classpath-multiple-bindings) bindings. Created #14 to examine the problem.
> 
> great work @emecas !

@mwithi effectively your commit  8ea01f30b  under #14 removes the warning, do you want me to add the commit here?

I'm still seeing one additional warning, but that is because we are using a too old SpringBoot version -->  `spring-boot-test-1.4.0.RELEASE` which resolves as `mockito-core-1.10.19` 

     WARNING: Illegal reflective access by org.mockito.cglib.core.ReflectUtils$2 (file:/.m2/repository/org/mockito/mockito-core/1.10.19/mockito-core-1.10.19.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)

**Comment by mwithi** [comment]
> @mwithi .. do you want me to add the commit here?
No, no need to do that, because it's just a warning and not related to tests by general, that's why I created a new PR for that.


> I'm still seeing one additional warning, but that is because we are using a too old SpringBoot version --> `spring-boot-test-1.4.0.RELEASE` which resolves as `mockito-core-1.10.19`
Where do you see this warning? Anyway, accordinf t

**Comment by emecas** [comment]
> 
> 
> > @mwithi .. do you want me to add the commit here?
> 
> No, no need to do that, because it's just a warning and not related to tests by general, that's why I created a new PR for that.
> 
> > I'm still seeing one additional warning, but that is because we are using a too old SpringBoot version --> `spring-boot-test-1.4.0.RELEASE` which resolves as `mockito-core-1.10.19`
> 
> Where do you see this warning? Anyway, according to the class logged it relates to this PR so it could be solved (if needed) under the latter itself

@mwithi  

I see the warning while running the patien test suite, but I could  reproduce that ` WARNING: Illegal reflective access` only in one on my machines (linux), in windows it is not happening, in case it persists and I can figure out a cause, I will make available a fix.

So, what is next for finally merge this PR? are we still waiting for @averni review?

thanks

**Comment by mwithi** [comment]
> So, what is next for finally merge this PR? are we still waiting for @averni review?

Yes

> I see the warning while running the patien test suite, but I could reproduce that WARNING: Illegal reflective access only in one on my machines (linux)

Very strange indeed...

**Pull request merged**